### DISCUSSION TITLE
如何让 xmake 将 exe 依赖的 dll 复制到 exe 所在目录？

### DISCUSSION THREAD

**Discussion by lc-soft** [xmake-io/xmake#1815] [Q&A] [answered]
跑 `xmake install -P . -o install` 后，bin 目录里并没有 dll，是需要手动写脚本复制这些 dll，还是我的 xmake.lua 配置有问题？

![image](https://user-images.githubusercontent.com/1730073/140640795-c7d2a54f-302b-4f76-bf23-ea2e28cbb7ad.png)

**Comment by waruqi** [comment]
依赖的shared target和 package 的dll 都会被安装

https://github.com/xmake-io/xmake/blob/736dc4db6368fa4dd1770e768c5577bc8d55255d/xmake/modules/target/action/install/windows.lua#L83

如果不行，提供个可复现的demo

**Reply by waruqi** [reply]
我搞了个demo试了下，没啥问题，dll可以正常 copy

**Reply by lc-soft** [reply]
@waruqi 试试？

https://github.com/lc-soft/xmake-target-install-test

**Reply by waruqi** [reply]
https://github.com/lc-soft/xmake-target-install-test/blob/0fa5f1107849616c0d125611becc8027a02a005f/libadd/test/xmake.lua#L3

你默认 add_requires 引用的还是静态库配置，尽管 target 本身是 shared 的，但是目前是否shared 是由用户 add_requires("libadd", {configs = {shared=true}}) 控制的

xmake 根据 add_requires 的配置选取 static/shared 包，然后包定义 on_install 根据 package:config("shared") 判断是否编译安装静态库还是动态库版本。

如果这个包只能 shared ，理论上只允许配置成 

add_requires("libadd", {configs = {shared=true}})

当然 xmake package 这块也可以改进 添加 shared readonly=true 配置，来固定shared配置包

**Reply by lc-soft** [reply]
@waruqi 改了，但仍然没效果

https://github.com/lc-soft/xmake-target-install-test/commit/4500e6cda6e437b801b3599151be03eff619e46e

**Comment by waruqi** [comment, marked as answer]
这是由于 xmake package 打的本地包里面缺少 libfiles配置导致，我改了下 再试试
